src/tls/ffi.c

```diff
@@ -0,0 +1,428 @@
+    close(r->fd_err_new);
+    result = unlink(r->err_name);
+    if (result != 0) {
+        fprintf(stderr, "unlink %s failed errno=%d ", r->out_name, errno); perror("failed");
+    }
+    free(r->err_name);
+#else
+    r->fd_out_new = -1;
+    *outmsg = NULL;
+    *errmsg = NULL;
+#endif
+}
+
+
+void redirect_stdio(/* out */ mitls_redirect *r)
+{
+#if REDIRECT_STDIO
+    int fdout;
+    int fderr;
+    static const char template[] = "/mitlscurlXXXXXX";
+    char *tmp;
+    char *out_name;
+    char *err_name;
+    size_t tmp_len;
+    
+    // Assume failure
+    r->fd_out_new = -1;
+    r->fd_err_new = -1;
+    
+    fflush(stdout);
+    fflush(stderr);
+    
+    tmp = getenv("TMP");
+    if (!tmp) {
+        tmp = getenv("TEMP");
+        if (!tmp) {
+            tmp = ".";
+        }
+    }
+    
+    tmp_len = strlen(tmp);
+    out_name = (char*)malloc(tmp_len + sizeof(template));
+    if (!out_name) {
+        return;
+    }
+    strcpy(out_name, tmp);
+    strcpy(out_name+tmp_len, template);
+    
+    err_name = strdup(out_name);
+    if (!err_name) {
+        free(out_name);
+        return;
+    }
+
+    // Create temp files
+    fdout = mkstemp(out_name);
+    if (fdout == -1) {
+        free(out_name);
+        free(err_name);
+        return;
+    }
+    fderr = mkstemp(err_name);
+    if (fderr == -1) {
+        close(fdout);
+        free(out_name);
+        free(err_name);
+        return;
+    }
+    r->fd_out_new = fdout;
+    r->fd_err_new = fderr;
+    r->out_name = out_name;
+    r->err_name = err_name;
+    
+    // Capture current stdout/stderr
+    r->fd_out_old = dup(STDOUT_FILENO);
+    r->fd_err_old = dup(STDERR_FILENO);
+    
+    // Redirect stdout and stderr to the temp files
+    dup2(fdout, STDOUT_FILENO);
+    dup2(fderr, STDERR_FILENO);
+#else
+    r->fd_out_new = -1;
+    r->fd_err_new = -1;
+#endif    
+}
+
+// Called by the host app to configure miTLS ahead of creating a connection
+int FFI_mitls_configure(mitls_state **state, const char *tls_version, const char *host_name, char **outmsg, char **errmsg)
+{
+    CAMLparam0();
+    CAMLlocal3(config, version, host);
+    int ret = 0;
+    mitls_redirect r;
+
+    *state = NULL;
+    *outmsg = NULL;
+    *errmsg = NULL;
+    redirect_stdio(&r);
+    
+    version = caml_copy_string(tls_version);  
+    host = caml_copy_string(host_name);
+    config = caml_callback2_exn(*g_mitls_FFI_Config, version, host);
+    if (Is_exception_result(config)) {
+        printf("Exception!  %s\n", caml_format_exception(Extract_exception(config)));
+    } else {
+        mitls_state * s;
+        
+        // Allocate space on the heap, to store an OCaml value
+        s = (mitls_state*)malloc(sizeof(mitls_state));
+        if (s) {
+            // Tell the OCaml GC about the heap address, so it is treated
+            // as a GC root, keeping the config object live.
+            s->fstar_state = config; 
+            caml_register_generational_global_root(&s->fstar_state);
```

Comment:
Rule 4 of http://caml.inria.fr/pub/docs/manual-ocaml-4.00/manual033.html says `caml_register_global_root`... the invariant that we must respect when using the `generational` version (documented at https://github.com/ocaml/ocaml/blob/trunk/byterun/caml/memory.h#L429) is fairly subtle... the invariant is respected in your code, though. Do you think there is a substantial performance gain from using this variant?